### PR TITLE
Limit number of events stored in TestEventListener

### PR DESCRIPTION
**Pull request by Mohit-Chakraborty**
In some cases, tests may spit out enough events to cause an OOM exception.
Usually, this is a sign of a rogue test. However, we can also harden the code in `TestEventListener` by dropping events beyond a certain capacity.

### CODE REVIEW

**Comment by Mohit-Chakraborty** [comment]
This was an issue with a test in the `Search` library. We have [fixed the test](https://github.com/Azure/azure-sdk-for-net/pull/21988) to not have the condition that caused it to log a very large number of error events.
I have created a separate PR for the listener to show that this change is not needed to fix the behavior of the test.

File: sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs [hunk 1]
```diff
@@ -14,6 +14,8 @@ public class TestEventListener : EventListener
     {
         private volatile bool _disposed;
         private readonly ConcurrentQueue<EventWrittenEventArgs> _events = new ConcurrentQueue<EventWrittenEventArgs>();
+        private uint _maxEventCount;
+        private const uint DefaultMaxEventCount = 100;
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs:18]: This seems very low.

Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs:18]: 1000?

Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs:18]: Would a level/message filter be a better solution?

Comment (Mohit-Chakraborty) [reply sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs:18]: I looked at the usage of the test event listener. Seems like it is used in a very focused manner where the test code expects around a dozen events and analyzes them. We can change to 1000, though I think that's an overkill.
Level or event-id based filter would not address the root of the issue, which is the unbounded growth of this DS. For example, in the case I looked at, we were logging error level events for the one particular id we were interested in.

Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs:18]: I'm worried about making tests flaky and hard to debug by silently dropping messaged. I'd rather throw and have an opt in.

**Review by pakrym — APPROVED** [review]

**Comment by pakrym** [comment]
Does it work for you in this shape?

**Comment by Mohit-Chakraborty** [comment]
> Does it work for you in this shape?

I think it's good. I have fixed the issue in our test to not generate excessive events. For the next unfortunate soul, it would be better to get this exception than an OOM that happens long after the offending test has passed.

**Comment by pakrym** [comment]
Nice find

File: sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs [hunk 2]
```diff
@@ -27,12 +29,25 @@ protected override void OnEventWritten(EventWrittenEventArgs eventData)
...
                 // Make sure we can format the event
                 EventSourceEventFormatting.Format(eventData);
                 _events.Enqueue(eventData);
             }
         }
 
+        public TestEventListener() : this(DefaultMaxEventCount)
+        { }
+
+        public TestEventListener(uint maxEventCount)
```
Comment (christothes) [inline sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs:46]: Let's add a comment here explaining what `maxEventCount` does.

**Review by christothes — APPROVED** [review]

**Pull request merged**